File: Edit_distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int l1 = word1.size();
        int l2 = word2.size();
        
        vector<int> dp(l2+1, 0);
        for(int i = 1; i <= l2; i++)
            dp[i] = i;
        
        for(int i = 1; i <= l1; i++)
        {
            int prev = i;
            for(int j = 1; j <= l2; j++)
            {
                int cur;
                if(word1[i-1] == word2[j-1])
                {
                    cur = dp[j-1];
                }
                else
                {
                    cur = min(min(dp[j-1],prev),dp[j])+1;
                }
                dp[j-1] = prev;
                prev =cur;
            }
            dp[l2] = prev;
        }
        return dp[l2];
    }
};

File: leetcode/dp/Edit Distance.go
func minDistance(word1 string, word2 string) int {
    m, n := len(word1), len(word2)
    dp := make([][]int, m+1)

    for i := 0; i < len(dp); i++{
        dp[i] = make([]int, n+1)
    }
    
    for i := 0; i < len(dp[0]); i++{
        dp[0][i] = i
    }

    for i := 0; i < len(dp); i++{
        dp[i][0] = i
    }
    for i := 1; i < len(dp); i++{
        for j := 1; j < len(dp[0]); j++{
            dp[i][j] =1+ min(dp[i-1][j], dp[i][j-1])
            if word1[i-1] == word2[j-1]{
                dp[i][j] = min(dp[i][j], dp[i-1][j-1])
            }else{
                dp[i][j] = min(dp[i][j], 1+dp[i-1][j-1])
            }
        }  
    }
    return dp[m][n]
}


func min(a int, b int) int{
    if a < b{
        return a
    }
    return b
}